### PR TITLE
Guidelines for Getting Started Guide contributions

### PR DESCRIPTION
**Pull request by erictune**

### CODE REVIEW

**Comment by erictune** [comment]
The first two commits are from #5111.  Please comment on those in that PR.

Just looking for comments on the last commit.

**Comment by erictune** [comment]
If this PR is merged, I propose it would immediately affect new commits of guides, but not immediately change existing guides.  I'd work with their authors to decide how to classify them and to test them when they are updated to a new binary version.

**Comment by erictune** [comment]
Oncall: This needs to be assigned a reviewer please.

**Comment by nikhiljindal** [comment]
Any suggestions for a suitable reviewer?
@thockin @brendandburns @alex-mohr ?

**Comment by erictune** [comment]
Assigned to @thockin.
FYI @craigmcl

**Comment by erictune** [comment]
@erickhan

File: docs/devel/writing-a-getting-started-guide.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+## Versioned Distro Guidelines
+These guidelines say *what* to do.  See the Rationale section for *why*.
+ - Send us a PR.
+ - Put the instructions go in `docs/getting-started-guides/...`. Scripts go there too.
+ - We may ask that you host binary assets or large amounts of code in our `contrib` directory or on your
+   own repo.
+ - Setup a cluster and run the [conformance test](../../docs/devel/conformance-test.md) against it, and report the
+   results in your PR.
+ - Add or update a row in [The Matrix](../../docs/getting-started-guides/README.md).
+ - State the binary version of kubernetes that you tested clearly in your Guide doc and in The Matrix.
```
Comment (thockin) [inline docs/devel/writing-a-getting-started-guide.md]: Also version of OS and provisioning tools?

**Comment by thockin** [comment]
LGTM, just one Q

File: docs/devel/writing-a-getting-started-guide.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Writing a Getting Started Guide
+This page gives some advice for anyone planning to write or update a Getting Started Guide for Kubernetes.
+It also gives some guidelines which reviewers should follow when reviewing a pull request for a
+guide. 
+
+A Getting Started Guide is instructions on how to create a Kubernetes cluster on top of a particular
+type(s) of infrastructure.  Infrastructure includes: the IaaS provider for VMs
+; the node OS; inter-node networking; and node Configuration Management system.
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: minor style nit: ';' should probably be on previous line

File: docs/devel/writing-a-getting-started-guide.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+A guide refers to scripts, Configuration Manangement files, and/or binary assets such as RPMs.  We call
+the combination of all these things needed to run on a particular type of infrastructure a
+**distro**.
+
+[The Matrix](../../docs/getting-started-guides/README.md) lists the distros.  If there is already a guide
+which is similar to the one you have planned, consider improving that one.
+
+
+Distros fall into two categories:  
+  - **development distros** are tested work with the latest Kubernetes source code.  But, there are
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: nit: the order of categories here should maybe match the order below? (right now they don't match)

File: docs/devel/writing-a-getting-started-guide.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  - **versioned distros** are tested to work with a particular binary release of Kubernetes.  These
+    come in a wide variety, reflecting a wide range of ideas and preferences in how to run a cluster.
+
+There are different guidelines for each.
+
+## Versioned Distro Guidelines
+These guidelines say *what* to do.  See the Rationale section for *why*.
+ - Send us a PR.
+ - Put the instructions go in `docs/getting-started-guides/...`. Scripts go there too.
+ - We may ask that you host binary assets or large amounts of code in our `contrib` directory or on your
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: Remove first "go" (after instructions)

File: docs/devel/writing-a-getting-started-guide.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    relatively few of these and the bar is much higher for creating one.
+  - **versioned distros** are tested to work with a particular binary release of Kubernetes.  These
+    come in a wide variety, reflecting a wide range of ideas and preferences in how to run a cluster.
+
+There are different guidelines for each.
+
+## Versioned Distro Guidelines
+These guidelines say *what* to do.  See the Rationale section for *why*.
+ - Send us a PR.
+ - Put the instructions go in `docs/getting-started-guides/...`. Scripts go there too.
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: Remove first "go" (after "instructions")

File: docs/devel/writing-a-getting-started-guide.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   own repo.
+ - Setup a cluster and run the [conformance test](../../docs/devel/conformance-test.md) against it, and report the
+   results in your PR.
+ - Add or update a row in [The Matrix](../../docs/getting-started-guides/README.md).
+ - State the binary version of kubernetes that you tested clearly in your Guide doc and in The Matrix.
+ - Even if you are just updating the binary version used, please still do a conformance test.
+     - If it worked before and now fails, you can ask on IRC, 
+       check the release notes since your last tested version, or look at git -logs for files in other distros
+       that are updated to the new version.
+ - versioned distros should typically not modify or add code in `cluster/`.  That is just scripts for developer
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: minor nit: capitalize versioned

File: docs/devel/writing-a-getting-started-guide.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+   distros.  
+ - If a versioned distro has not been updated for many binary releases, it may be dropped frome the Matrix.
+
+If you have a cluster partially working, but doing all the above steps seems like too much work,
+we still want to hear from you.  We suggest you write a blog post or a Gist, and we will link to it on our wiki page.
+Just file an issue or chat us on IRC and one of the committers will link to it from the wiki.
+
+## Development Distro Guidelines
+These guidelines say *what* to do.  See the Rationale section for *why*.
+  - the main reason to add a new development distro is to support a new IaaS provider (VM and
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: minor nit: capitalization in this list is very inconsistent

File: docs/devel/writing-a-getting-started-guide.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    network management).  This means implementing a new `pkg/cloudprovider/$IAAS_NAME`.  
+  - Development distros should use Saltstack for Configuration Management.
+  - development distros need to support automated cluster creation, deletion, upgrading, etc.
+    This mean writing scripts in `cluster/$IAAS_NAME`.
+  - all commited to the tip of this repo need to not break any of the development distros
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: Missing noun between "all" and "committed" - probably want "code" or "changes" or something like that? Also "committed" is misspelled.

File: docs/devel/writing-a-getting-started-guide.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    - the author is responsible for making changes necessary on all the cloud-providers if the
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: Should clarify that this means author of the change, not author who added the distro

File: docs/devel/writing-a-getting-started-guide.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      change affects any of them, and reverting the change if it breaks any of the CIs.
+  - a development distro needs to have an organization which owns it.  This organization needs to: 
+    - Setting up and maintaining Continuous Integration testing on all commits against their IaaS.
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: Are we going to require that CI results are publicly available?

How frequently do CIs need to run tests? Running the full e2e suite on every commit is probably not feasible.

File: docs/devel/writing-a-getting-started-guide.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+   for consistency.
+ - We want users to have a uniform experience with Kubernetes whenever they follow instructions anywhere
+   in our Github repository.  So, we ask that versioned distros pass a **conformance test** to make sure
+   really work.
+ - We ask versioned distros to **clearly state a version**.  People pulling from Github may 
+   expect any instructions there to work at Head, so stuff that has not been tested at Head needs
+   to be called out.  We are still changing things really fast, and, while the REST API is versioned,
+   it is not practical at this point to version or limit changes that affect distros.  We still change
+   flags at the Kubernetes/Infrastructure interface.
+   so instructions can rot quickly.
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: minor nit: is this part of the previous sentence? If so, delete previous period and join the lines.

File: docs/devel/writing-a-getting-started-guide.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+ - We want to **limit the number of development distros** for several reasons.  Developers should
+   only have to change a limited number of places to add a *new feature.  Also, since we will
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: stray * before "new"

File: docs/devel/writing-a-getting-started-guide.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+   gate commits on passing CI for all distros, and since end-to-end tests are typically somewhat
+   flaky, it would be highly likely for there to be false positives and CI backlogs with many CI pipelines.
+ - We do not require versioned distros to do **CI** for several reasons.  It is a steep
+   learning curve to understand our our automated testing scripts.  And it is considerable effort
+   to fully automate setup and teardown of a cluster, which is needed for CI.  And, not everyone
+   has the time and money to run CI and to respond to We do not want to
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: "... and to respond to" sentence is cut off.

File: docs/devel/writing-a-getting-started-guide.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+   discourage people from writing and sharing guides because of this.  
+ - Versioned distro authors are free to run their own CI and let us know if there is breakage, but we
+   will not include them as commit hooks -- there cannot be so many commit checks that it is impossible
+   to pass them all.
+ - We require a single Configuration Management tool for development distros.  If there were more
+   than one, the core developers would have to learn multiple tools and update config in multiple
+   places.  **Saltstack** happens to be the one we picked when we started the project.  We
+   welcome versioned distros that use any tool; there are already examples of 
+   CoreOS Fleet, Ansible, and others.
+   This is a necessary bar for code that is subject to CI testing, but would block
```
Comment (ixdy) [inline docs/devel/writing-a-getting-started-guide.md]: "but would block" sentence is cut off

File: docs/devel/writing-a-getting-started-guide.md (hunk 1 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: The main goal of stating the k8s version is so users and devs can track which guides are actively maintained.  One number accomplishes that.

The goal here isn't to to state all the prereqs -- the getting-started-guide should show that. In the future we might ask for more detail on what was testing, but I don't think we need to be that heavyweight yet.

File: docs/devel/writing-a-getting-started-guide.md (hunk 2 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 3 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 5 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 6 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 7 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: it makes sense to me.

File: docs/devel/writing-a-getting-started-guide.md (hunk 8 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 9 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 10 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: reworded.  Some of this stuff we will figure out as we go.

File: docs/devel/writing-a-getting-started-guide.md (hunk 11 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 12 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed

File: docs/devel/writing-a-getting-started-guide.md (hunk 13 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

File: docs/devel/writing-a-getting-started-guide.md (hunk 14 above)
Comment (erictune) [inline docs/devel/writing-a-getting-started-guide.md]: fixed.

**Comment by erictune** [comment]
@ixdy please merge if satisfied.

**Comment by ixdy** [comment]
LGTM

**Pull request merged**